### PR TITLE
Upgraded to 3.9.1

### PR DESCRIPTION
**Pull request by nataliecarey**
This includes the new features and some old features which we hadn't included before.

This specifically excludes some examples which are not reasonably compatible with our strong typed port.  We are in discussion with GDS about those.

### CODE REVIEW

File: src/main/scala/uk/gov/hmrc/govukfrontend/views/viewmodels/CommonJsonFormats.scala [hunk 1]
```diff
@@ -56,4 +57,60 @@ object CommonJsonFormats {
     override def writes(o: Option[Html]): JsObject =
       o.map(o => Json.obj("conditional" -> Json.obj("html" -> o.body))).getOrElse(Json.obj())
   }
+
+  val attributesReads: Reads[Map[String, String]] = new Reads[Map[String, String]] {
```
Comment (matthewmascord) [inline src/main/scala/uk/gov/hmrc/govukfrontend/views/viewmodels/CommonJsonFormats.scala:61]: Would it be worth converting these to anonymous functions i.e.
```scala
val attributesReads: Reads[Map[String, String]] = (json: JsValue) => {
    val keyValueTuples = json.as[JsObject].keys.map { key =>
      val maybeValue: Option[String] = (json \ key).asOpt[String].orElse {
        (json \ key).asOpt[Int].map(_.toString).orElse {
          (json \ key).asOpt[Boolean].map(_.toString)
        }
      }
      maybeValue.map(v => (key, v))
    }
    JsSuccess(keyValueTuples.flatten.toMap)
  }
```

File: src/main/scala/uk/gov/hmrc/govukfrontend/views/viewmodels/CommonJsonFormats.scala [hunk 2]
```diff
@@ -56,4 +57,60 @@ object CommonJsonFormats {
...
+          json.asOpt[Boolean].map(_.toString)
+        }
+      }
+      JsSuccess(maybeString)
+    }
+  }
+
+  val forgivingStringReads: Reads[String] = new Reads[String] {
+    override def reads(json: JsValue): JsResult[String] = {
+      val maybeString = json.asOpt[String].orElse {
```
Comment (matthewmascord) [inline src/main/scala/uk/gov/hmrc/govukfrontend/views/viewmodels/CommonJsonFormats.scala]: There is a little duplication of the code from line 77 - could it be factored out?

File: src/main/scala/uk/gov/hmrc/govukfrontend/views/viewmodels/input/Input.scala [hunk 3]
```diff
@@ -81,7 +85,9 @@ object Input extends JsonDefaultValueFormatter[Input] {
         (__ \ "autocomplete").writeNullable[String] and
         (__ \ "pattern").writeNullable[String] and
         (__ \ "attributes").write[Map[String, String]] and
-        (__ \ "spellcheck").writeNullable[Boolean]
+        (__ \ "spellcheck").writeNullable[Boolean] and
+        (__ \ "prefix").writeNullable[PrefixOrSuffix] and  //TODO: Make sure this actually writes (write a test for it)
```
Comment (matthewmascord) [inline src/main/scala/uk/gov/hmrc/govukfrontend/views/viewmodels/input/Input.scala]: Is this still TODO?

File: src/main/scala/uk/gov/hmrc/govukfrontend/views/viewmodels/CommonJsonFormats.scala (hunk 2 above)
Comment (JoPintoPaul) [reply src/main/scala/uk/gov/hmrc/govukfrontend/views/viewmodels/CommonJsonFormats.scala]: Cleaned up and pushed to branch.

File: src/main/scala/uk/gov/hmrc/govukfrontend/views/viewmodels/CommonJsonFormats.scala (hunk 1 above)
Comment (JoPintoPaul) [reply src/main/scala/uk/gov/hmrc/govukfrontend/views/viewmodels/CommonJsonFormats.scala:61]: This is a good question. To me, the longer code is more explanatory, and better follows the documentation for creating custom JSON readers from the Play docs:
```
new Reads[T] {
  def reads(json: JsValue): JsResult[T] {
    ...
  }
}
```
The syntax you suggest is less verbose, but I do feel it loses some of readability / understand-ability.

**Review by matthewmascord — APPROVED** [review]

**Pull request merged**